libsupport/src/Plugin.cpp

```diff
@@ -0,0 +1,187 @@
+#include "katana/Plugin.h"
+
+#include <dlfcn.h>
+
+#include <cstdlib>
+#include <unordered_map>
+
+#include <boost/filesystem.hpp>
+#include <katana/Logging.h>
+#include <katana/Result.h>
+#include <katana/Strings.h>
+
+namespace {
+
+class PluginLoader {
+  std::unordered_map<std::string, katana::Plugin> plugins_;
+
+public:
+  void LoadPlugin(const boost::filesystem::path& path) {
+    if (plugins_.count(path.string())) {
+      return;
+    }
+
+    void* handle = dlopen(path.c_str(), RTLD_LAZY | RTLD_LOCAL);
+    if (!handle) {
+      KATANA_LOG_WARN(
+          "skipping plugin {}: failed to dynamically link plugin: {}", path,
+          katana::ResultErrno());
+      return;
+    }
+    const auto KatanaPluginInit =
+        reinterpret_cast<const katana::PluginMetadata* (*)()>(
+            dlsym(handle, "KatanaPluginInit"));
+    const katana::PluginMetadata* metadata = KatanaPluginInit();
+    if (!metadata) {
+      KATANA_LOG_WARN("plugin {} may have failed to load", path);
+    }
+    KATANA_LOG_VERBOSE("Loaded plugin: {}", path);
+    plugins_.emplace(
+        path.string(), katana::Plugin{
+                           .metadata = metadata,
+                           .so_path = path.string(),
+                           .so_handle = handle,
+                       });
+  }
+
+  void LoadAllPlugins() {
+    if (!plugins_.empty()) {
+      return;
+    }
+
+    auto path = katana::GetPluginPath();
+    for (const auto& p : path) {
+      KATANA_LOG_VERBOSE("Loading plugins from: {}", p);
```

Comment:
OK. I thought error messages where initial lowercase and log messages where initial capitol because they were sentences. I'll change this and in other places.

Reply:
I actually don't know if we have a convention for log messages. We should probably pick one. I was more reacting to the varying capitialization in the log, warn and warn_once messages in this file.